python/TestHarness/schedulers/Scheduler.py

```diff
@@ -0,0 +1,539 @@
+        # The tester pool and status pool can exhange jobs between each other but it is always the status pool
+        # that will have items in the queue last (that last job with a finished status to be printed)
+
+        # Close the tester_pool so jobs can no longer be scheduled for launching
+        self.tester_pool.close()
+
+        # Wait for the tester_pool to empty
+        self.tester_pool.join()
+
+        # Close the status_pool so jobs can no longer be scheduled for status printing
+        self.status_pool.close()
+
+        # Wait for the status_pool to empty
+        self.status_pool.join()
+
+    # Create thread timers for various timing tasks (timeouts, long running)
+    # TODO:  to clever? separate these? The idea was we can add as many different timers as we want
+    #        and the only thing that would change would be this method.
+    def handleTimers(self, start_timers=None, stop_timers=None):
+        # Start some timers
+        if start_timers:
+            tester = start_timers.getTester()
+
+            # Long running timer (RUNNING...)
+            long_running_timer = threading.Timer(float(tester.getMinReportTime()),
+                                                 self.handleLongRunningJobs,
+                                                 (start_timers,))
+            long_running_timer.start()
+
+            # Failed TIMEOUT timer
+            timeout_timer = threading.Timer(float(tester.getMaxTime()),
+                                          self.handleTimeoutJobs,
+                                          (start_timers,))
+            timeout_timer.start()
+
+            return (long_running_timer, timeout_timer)
+
+        # Stop any timers we started
+        elif stop_timers:
+            for timer in stop_timers:
+                timer.cancel()
+
+    # Handle long running jobs
+    def handleLongRunningJobs(self, job_container):
+        tester = job_container.getTester()
+        tester.specs.addParam('caveats', ['FINISHED'], "")
+        tester.setStatus('RUNNING...', tester.bucket_pending)
+        self.queueJobs(status_job_containers=[job_container])
+
+    # Handle jobs that are timing out
+    def handleTimeoutJobs(self, job_container):
+        tester = job_container.getTester()
+        tester.setStatus('TIMEOUT', tester.bucket_fail)
+        job_container.killProcess()
+
+    # return load average
+    def getLoad(self):
+        loadAverage = 0.0
+        try:
+            loadAverage = os.getloadavg()[0]
+        except AttributeError:
+            pass      # getloadavg() not available in this implementation of os
+        return loadAverage
+
+    def satisfyLoad(self):
+        # We'll always run at least one job regardless of load or we'll starve!
+        while self.workers_in_use > 1 and self.getLoad() >= self.average_load:
+            sleep(1.0)
+
+    # Return bool if we have enough resouces to run the job
+    def checkAvailableSlots(self, job_container):
```

Comment:
`reserveSlots()` is probably a better name. This method will block and more importantly it actually reserves the slots so "check" isn't the right name.